### ISSUE TITLE
japicmp fails generating a report for a scala project

### ISSUE THREAD

**Issue by rmetzger** [siom79/japicmp#90]
I'm trying to use `japicmp` for [Apache Flink](http://flink.apache.org), to ensure interface stability with the upcoming 1.0 release.
Flink is a mixed Java/Scala project, and it seems that japicmp fails while generating the XML report:

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.6.2-SNAPSHOT:cmp (default) on project flink-scala: Execution default of goal com.github.siom79.japicmp:japicmp-maven-plugin:0.6.2-SNAPSHOT:cmp failed: XSLT transformation failed: javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: An invalid XML character (Unicode: 0x6) was found in the value of attribute "value" and element is "newElementValue". -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.6.2-SNAPSHOT:cmp (default) on project flink-scala: Execution default of goal com.github.siom79.japicmp:japicmp-maven-plugin:0.6.2-SNAPSHOT:cmp failed: XSLT transformation failed: javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: An invalid XML character (Unicode: 0x6) was found in the value of attribute "value" and element is "newElementValue".
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:224)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default of goal com.github.siom79.japicmp:japicmp-maven-plugin:0.6.2-SNAPSHOT:cmp failed: XSLT transformation failed: javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: An invalid XML character (Unicode: 0x6) was found in the value of attribute "value" and element is "newElementValue".
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	... 20 more
Caused by: japicmp.exception.JApiCmpException: XSLT transformation failed: javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: An invalid XML character (Unicode: 0x6) was found in the value of attribute "value" and element is "newElementValue".
	at japicmp.output.xml.XmlOutputGenerator.createXmlDocumentAndSchema(XmlOutputGenerator.java:164)
	at japicmp.output.xml.XmlOutputGenerator.generate(XmlOutputGenerator.java:60)
	at japicmp.maven.JApiCmpMojo.generateXmlOutput(JApiCmpMojo.java:297)
	at japicmp.maven.JApiCmpMojo.executeWithParameters(JApiCmpMojo.java:109)
	at japicmp.maven.JApiCmpMojo.execute(JApiCmpMojo.java:78)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	... 21 more
Caused by: javax.xml.transform.TransformerException: javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: An invalid XML character (Unicode: 0x6) was found in the value of attribute "value" and element is "newElementValue".
	at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.transform(TransformerImpl.java:758)
	at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.transform(TransformerImpl.java:359)
	at japicmp.output.xml.XmlOutputGenerator.createXmlDocumentAndSchema(XmlOutputGenerator.java:154)
	... 26 more
Caused by: javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: An invalid XML character (Unicode: 0x6) was found in the value of attribute "value" and element is "newElementValue".
	at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.getDOM(TransformerImpl.java:587)
	at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.transform(TransformerImpl.java:748)
	... 28 more
Caused by: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: An invalid XML character (Unicode: 0x6) was found in the value of attribute "value" and element is "newElementValue".
	at com.sun.org.apache.xalan.internal.xsltc.dom.XSLTCDTMManager.getDTM(XSLTCDTMManager.java:464)
	at com.sun.org.apache.xalan.internal.xsltc.dom.XSLTCDTMManager.getDTM(XSLTCDTMManager.java:252)
	at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.getDOM(TransformerImpl.java:565)
	... 29 more
[ERROR] 
```

I think adding a configuration flag for disabling the report would be a first good approach to resolve the issue (we don't really need a report. its good enough if the plugin is failing the build on breaking API changes). What do you think?

**Comment by siom79** [comment]
Wow, you have control characters like ACK (0x6) in your annotation values? Unfortunately this is a basic problem when using JAXB for XML marshalling (see for example [here](http://stackoverflow.com/questions/10121717/how-to-write-and-read-special-characters-and-symbols-from-xml-using-jaxb)).

Experimentally I have added an escaping of special characters for the XML attribute `value` of the element `newElementValue` as I can deduce from your error message. You can try it on the `develop` branch:
```
git clone https://github.com/siom79/japicmp.git
cd japicmp
git checkout develop
mvn install
```
Let me know what you think.

**Comment by rmetzger** [comment]
Thank you for addressing the issue so quickly!

I don't have these characters in my annotation values. 
I think the issue is that the Scala compiler is generating crazy bytecode and javaassist/japicmp/jaxb is not able to properly handle it.

I'll try out the development branch soon ...

**Comment by rmetzger** [comment]
Allright, I've tested it.
The good news is: The plugin is not failing anymore, it generates a report.

However, the report looks like this: https://transfer.sh/PwRaH/japicmp.html (the CSS is off due to transfer.sh, but the report is showing the value of byte arrays, which probably contain everything imaginable ;)
Maybe it would make sense not to show the value of byte and byte[] ?

**Comment by siom79** [comment]
Thanks for provding the report via transfer.sh. 

I am totally with you that it may not make sense to show byte[] values, but digging a bit deeper into that specific Scala annotation shows that it actually uses a `String` (see [here](http://grepcode.com/file/repo1.maven.org/maven2/org.scala-lang.virtualized/scala-library/2.11.2/scala/reflect/ScalaSignature.java/)) to encode additional information for Scala's reflection mechanism (see [here](http://www.scala-lang.org/old/sid/10)).

With this information a general solution gets a little bit more subtle. Maybe it makes sense to add an option that does not evaluate annotations or that even allows you to exclude specific annotations...

**Comment by siom79** [comment]
As a first step and because I have had this option on my todo list for some time, I have added a new option that allows you to disable the evaluation of annotations completely (see new issue [#91](https://github.com/siom79/japicmp/issues/91)).

**Comment by rmetzger** [comment]
I agree that disabling annotations / excluding specific annotations is a good approach to get rid of the issue.

Thank you for looking into this so closely.

**Comment by siom79** [comment]
Released with 0.6.2.
